src/validation.cpp

```diff
@@ -1090,6 +1090,33 @@ bool AcceptToMemoryPool(CTxMemPool& pool, TxValidationState &state, const CTrans
     return AcceptToMemoryPoolWithTime(chainparams, pool, state, tx, GetTime(), plTxnReplaced, bypass_limits, test_accept, fee_out);
 }
 
+// Calculate the fee of a transaction by looking up inputs from the mempool if
+// necessary. Return true if the fee was successfully calculated, or false if
+// an error with the transaction was discovered.
+// If the transaction is valid, then the fee returned is correct.
+// However, success does not imply that the transaction is valid.
+bool GetTransactionFee(CTxMemPool& pool, TxValidationState &state, const CTransactionRef &tx, CAmount *fee_out)
+{
+    LOCK2(cs_main, pool.cs);
+
+    CCoinsViewMemPool viewmempool(&::ChainstateActive().CoinsTip(), pool);
+    CCoinsViewCache view(&viewmempool);
+    view.GetBestBlock();
```

Comment:
What is this line for? `GetBestBlock()` is a const method, and the return value is being discarded here, so this appears to be a no-op.

Reply:
`GetBestBlock()` is const, but `hashBlock` is mutable, and what this function does is load the block hash associated with the utxo set into the `CoinsViewCache` object.  

The reason it is here is so that when we invoke `GetSpendHeight(m_view)` in the call to `Consensus::CheckTxInputs()`, we're doing so with the right height of the next block, which is necessary for the coinbase maturity check that happens in that function.

I think omitting this line would probably cause the code to crash, due to a null dereference in `GetSpendHeight()`?

Reply:
I tried removing the line and it looks okay. I might be reading this wrong but I think when we call CheckTxInputs` with the `GetSpendHeight(view)`:
```c
if (!Consensus::CheckTxInputs(tx, state, view, GetSpendHeight(view), *fee_out)) {
```

[`GetSpendHeight()`](https://github.com/bitcoin/bitcoin/blob/8aa3a4a498742dc42631956ed3599b3fb2b80302/src/validation.cpp#L1477) calls `GetBestBlock()` for `view` as well:

```c 
int GetSpendHeight(const CCoinsViewCache& inputs)
{
    LOCK(cs_main);
    CBlockIndex* pindexPrev = LookupBlockIndex(inputs.GetBestBlock());
    return pindexPrev->nHeight + 1;
}
```

Reply:
Good point!